### ISSUE TITLE
Class version of AUROC metric

### ISSUE THREAD

**Issue by NumesSanguis** [PyTorchLightning/pytorch-lightning#4547]
## 🚀 Feature

Class version of AUROC metric following the v1.x.x standard, so we can do:
```python
auroc = AUROC(num_classes=4)
auroc(pred, target)
auroc.compute()
```

### Motivation

Class based Metrics can be nicer to use in the PyTorch Lightning workflow, so why not add one for AUROC?

### Pitch

AUROC will automatically use the correct functional function, `auroc` or `multiclass_auroc`, based on whether `num_classes == 2` or `num_classes > 2`, respectively.
Not sure why `auroc` or `multiclass_auroc` are 2 separate functions?

### Alternatives

Implement the AUROC Metric yourself.

### Additional context

**Considerations:**
How to handle `update()`? While it's true that most likely the whole dataset will include all classes in `pred` and `targets`, this cannot be assumed for a single batch.

**Comment by NumesSanguis** [comment]
I've created a possible implementation of a Metric Class for AUROC:

```python
from typing import Optional, Any, Sequence
import torch
from pytorch_lightning.metrics import Metric
from pytorch_lightning.metrics.functional.classification import auroc, multiclass_auroc


class AUROC(Metric):
    def __init__(
            self,
            num_classes: int,
            sample_weight: Optional[Sequence] = None,
            compute_on_step: bool = False,  # True likely crashes if not every batch contains all classes
            dist_sync_on_step: bool = False,
            process_group: Optional[Any] = None,
    ):
        super().__init__(
            compute_on_step=compute_on_step,
            dist_sync_on_step=dist_sync_on_step,
            process_group=process_group,
        )

        self.num_classes = num_classes
        self.sample_weight = sample_weight
        self.add_state("all_preds", default=[])
        self.add_state("all_target", default=[])

    def update(self, preds: torch.Tensor, target: torch.Tensor) -> None:
        self.all_preds.append(preds)
        self.all_target.append(target)

    def compute(self):
        preds_tensor = torch.cat(self.all_preds)
        target_tensor = torch.cat(self.all_target)

        if self.num_classes == 2:
            return auroc(preds_tensor, target_tensor, sample_weight=self.sample_weight)
        else:
            return multiclass_auroc(preds_tensor, target_tensor, num_classes=self.num_classes,
                                    sample_weight=self.sample_weight)

```

## Notes & Concerns
- `compute_on_step` should almost always be `False`, otherwise it will fail on a batch that does not include all classes (See *Considerations* in the first post)
- This implementation stores all `preds` and `target`. This might be unwanted, but I couldn't think of a way that would not fail if in a batch `preds` and `target` does not contain all classes.
- `num_classes == 2` is treated differently from `num_classes > 2`, because PL has `auroc` and `multiclass_auroc`. Is it not possible for `multiclass_auroc` to also cover only 2 classes? Or is there a speed advantage or something ing `auroc`?

**Comment by tchaton** [comment]
Hey @NumesSanguis,

Thanks for this idea. I like it !
Would you mind creating a PR and ping @teddykoker.
Please, make sure to add tests :)

Best regards,
Thomas Chaton.

**Comment by SkafteNicki** [comment]
Hi @NumesSanguis 
Sounds great. I have just opened a PR for the three metric PrecisionRecall, ROC, AveragePrecision (https://github.com/PyTorchLightning/pytorch-lightning/pull/4549) that adds class interface for these. AUROC and ROC are closely connected so maybe you could base your PR on mine, such that the AUROC implementation just calls the new ROC implementation?

**Comment by NumesSanguis** [comment]
@SkafteNicki Great work! Then I'll wait till that PR has been merged.

**Comment by ddrevicky** [comment]
Hi @NumesSanguis good work :)

I can see the problem with not all classes being in a batch.
We could: 
- raise an error - that's exactly what`sklearn` does in that case. 
- issue a warning and not compute the metric for that batch (perhaps preferable?)
- do something like you're doing but we probably don't want to store all predictions and targets. Besides the memory issue, there is also a problem with the predictions becoming stale. If the dataset is very large and we are including predictions from the start of the epoch towards the epoch's end, then that might not reflect the model's actual performance. Perhaps it's not the responsibility of PL that the metric is always computed. Maybe it's user's responsibility to understand the requirements of the metric that they are using and if they want to have it computed it every batch then they should come up with a hack themselves but at least they will be aware that they are using one :). 
- do some other thing I have not thought of, perhaps @SkafteNicki can guide you better

Other thoughts:
- You can take a look at sklearn's [docs](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.roc_auc_score.html) and source for [auroc](https://github.com/scikit-learn/scikit-learn/blob/0fb307bf3/sklearn/metrics/_ranking.py#L248) I found both quite useful.
- the functional version of `multiclass_auroc` and `auroc` could be merged IMHO similarly as is done in [sklearn](https://github.com/scikit-learn/scikit-learn/blob/0fb307bf3/sklearn/metrics/_ranking.py#L248). Some dimension checks would be necessary to find out whether for the binary case, the user is passing a 1d array for positive class probs `(n_samples,)` or `(n_samples, 2)`. Again, you could take a look at `sklearn`'s version's parameters `y_true`, `y_scores`

**Comment by NumesSanguis** [comment]
Thank you for your thoughts @ddrevicky !

1. & 2. Unlike accuracy, AUROC is used to get a sense of the performance of each class. Therefore, it is often used for imbalanced datasets. Take for example [AudioSet](https://research.google.com/audioset/), an audio dataset with 527 classes, [which also uses AUC](https://research.google/pubs/pub45857/) as one of the metrics to determine performance. Since having a batch with all 527 every time is not reasonable, I don't think we can take `sklearn`'s approach on a batch level. A warning/not computing is also not feasible in this example (you would get none or only a few batches of calculations).
3. Yes, something not storing all predictions and targets would be preferred. Is it becoming stale really an issue? If you would use `sklearn` or the functional version like `multiclass_auroc`, you would likely also only call it at the end of an epoch right?

-----
1. Thanks for the links! What should we do with `'ovr'` and `'ovo'`? Anyone having a recommendation for this?
2. I think that would make the functional `auroc` function more straightforward to use. As long as we force the user to specify `num_classes` in the class based AUROC, this is not an issue for this Issue though.

**Comment by ddrevicky** [comment]
@NumesSanguis 
3. You're right staleness is not an issue if validation is done at epoch's end, I got confused a little :)
-------
1. We only have the one-vs-rest (`'ovr'`) version implemented and that's what's being [tested against in sklearn](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/tests/metrics/functional/test_classification.py#L340), so no need for the user to specify that. It could be mentioned in the docs though.
2. I agree with you

**Comment by SkafteNicki** [comment]
Closed by #5479
